### DISCUSSION TITLE
How do I create Glyphs that scale when zooming to always appear as the same size, similar to point and line actors?

### DISCUSSION THREAD

**Discussion by danelinquist** [pyvista/pyvista#5205] [Q&A] [answered]
I'm developing an FE post processor using PyVista and want to display the magnitudes of vectorized data (e.g., displacements, strains, etc...) as arrows situated at the center of each element, similar to the third image in this example: https://docs.pyvista.org/version/stable/examples/01-filter/glyphs.html#sphx-glr-examples-01-filter-glyphs-py
![image](https://github.com/pyvista/pyvista/assets/71725627/f77ee928-7da0-4d54-b97b-e78d0b153628)

This example gives me most of what I need in regards to mapping the data values to scale the length & color of the arrows, however, to avoid occluding the results I need the arrows to stay the same relative size regardless of how far zoomed in I am. Point, line, and label actors appear to have the functionality that I am after, but its unclear to me how this is handled through PyVista/VTK.

Are there other examples or documentation that someone could point me to to accomplish this?

Thanks,

Dane.

**Comment by darikg** [comment, marked as answer]
You can use [VTK algorithm techniques](https://docs.pyvista.org/version/stable/examples/99-advanced/plotting-algorithms.html) to have the glyph's scale factor change dynamically with camera distance. Here's my attempt, adapting from this [vtkDistanceToCamera](https://examples.vtk.org/site/Cxx/Visualization/DistanceToCamera/) example

```python
import numpy as np
import pyvista as pv
from vtkmodules.vtkRenderingCore import vtkDistanceToCamera
from vtkmodules.vtkFiltersCore import vtkGlyph3D

sphere = pv.Sphere(radius=3.14)
sphere["direction"] = 0.3 * np.vstack((
   np.sin(sphere.points[:, 0]),
   np.cos(sphere.points[:, 1]),
   np.cos(sphere.points[:, 2]),
)).T

distanceToCamera = vtkDistanceToCamera()
distanceToCamera.SetInputData(sphere)
distanceToCamera.SetScreenSize(100.0)

glyph = vtkGlyph3D()
glyph.SetSourceData(pv.Arrow())
glyph.SetInputConnection(distanceToCamera.GetOutputPort())
glyph.SetScaleFactor(0.5)

# From the vtkGlyph3d docs: You can set what arrays to use for the
# scalars, vectors, normals, and color scalars by using the
# SetInputArrayToProcess methods in vtkAlgorithm. The first array is
# scalars, the next vectors, the next normals and finally color scalars.
glyph.SetScaleModeToScaleByScalar()
glyph.SetInputArrayToProcess(0, 0, 0, pv.FieldAssociation.POINT.value, "DistanceToCamera")
glyph.SetVectorModeToUseVector()
glyph.SetInputArrayToProcess(1, 0, 0, pv.FieldAssociation.POINT.value, "direction")
glyph.OrientOn()
glyph.SetColorModeToColorByVector()
glyph.SetInputArrayToProcess(3, 0, 0, pv.FieldAssociation.POINT.value, "direction")

pl = pv.Plotter()
distanceToCamera.SetRenderer(pl.renderer)
distanceToCamera.Update()
pl.add_mesh(sphere, color='grey', opacity=.3)
pl.add_mesh(glyph)

pl.show()
```

**Reply by danelinquist** [reply]
Thanks Darkig! This seems to do what I want, and should work with customized objects as well. 
Something I notice is it runs noticeably slower than the original example, and takes a couple seconds to react to zooms & pans. My guess is there is some background calculations that have to occur for each arrow to scale accordingly, but I'm unsure where exactly this is introduced. Do you know if this is side effect of the added vtkDistanceToCamera algorithm, and if so is there anything I can add or a different way of managing the scaling that could be done?

**Reply by darikg** [reply]
Even the example code there exhibits that behavior, or only your specific application?

**Reply by danelinquist** [reply]
Apologies, I should have been more specific. Your example case runs perfectly fine speed wise, and after digging into what I was trying to do, I believe I fixed the speed issues I was having.
The other cases I am trying to set up require different symbols than the PyVista arrow. If I load in our desired symbol from an STL mesh, and convert it to PolyData, I am able to drop it into the SetSourceData call instead of the arrow (code excerpt below).
```
# Try and load in an stl...
reader = pv.get_reader('HTC_Glyph.stl')
glyphMesh = reader.read()
HTCGlyph = pv.PolyData(glyphMesh)

glyph = vtkGlyph3D()
glyph.SetSourceData(HTCGlyph)
glyph.SetInputConnection(distanceToCamera.GetOutputPort())
glyph.SetScaleFactor(0.002)
```
This works as expected display wise, and after setting the scale factor accordingly, I get something like this:

![image](https://github.com/pyvista/pyvista/assets/71725627/251d1908-1a83-4086-83b2-30119eda2cb2)

Display wise, this was very close to what I am after, it just ran extremely slow. Digging into it further, I found the STL I was using had several thousand elements in it. Using the decimate function in PyVista, I was able to reduce the element count to a couple hundred elements which runs much faster.

A couple additional questions for you: 
1.) Does this seem like a valid approach for implementing customized glyph geometry (STL to PolyData), or should I be using a different method? 
2.) The other thing I would like to be able to modify, is the density of the glyphs being displayed relative to the camera distance. In this example the glyphs are fixed at each node of the sphere. Ideally, for my case the number of glyphs would reduce as the user zooms out, becoming more sparse so as not to occlude the underlaying mesh. Is there a method to thin the glyph field (e.g., insert a glyph every nth node or similar) that could also be tied in to the vtkDistanceToCamera algorithm? I'm shooting for something like this:

![image](https://github.com/pyvista/pyvista/assets/71725627/1e0e1c5a-b987-46c6-b0a8-f08fb154c4f3)

**Reply by darikg** [reply]
> Digging into it further, I found the STL I was using had several thousand elements in it. Using the decimate function in PyVista, I was able to reduce the element count to a couple hundred elements which runs much faster.

Nice!

> The other thing I would like to be able to modify, is the density of the glyphs being displayed relative to the camera distance. In this example the glyphs are fixed at each node of the sphere. Ideally, for my case the number of glyphs would reduce as the user zooms out, becoming more sparse so as not to occlude the underlaying mesh. Is there a method to thin the glyph field (e.g., insert a glyph every nth node or similar) that could also be tied in to the vtkDistanceToCamera algorithm?

That's a neat idea. Here's my first stab at it -- it interjects a clean filter between the sphere and the distanceToCamera filter, 
and updates the point merging tolerance on zoom events:

```python
from vtkmodules.vtkFiltersCore import vtkGlyph3D, vtkStaticCleanPolyData

clean = vtkStaticCleanPolyData()
clean.SetInputData(sphere)
clean.SetToleranceIsAbsolute(False)  # Use fraction of sphere bounding box
clean.SetTolerance(0)  # No merging by default. See also SetAbsoluteTolerance

distanceToCamera = vtkDistanceToCamera()
distanceToCamera.SetInputConnection(clean.GetOutputPort())  # Instead of .SetInputData(sphere)

... # Same as the example above

def on_zoom(*_):
    d = np.linalg.norm(pl.camera.position)  # distance from origin
    tol = np.clip((d - 3.14) / 10, 0, 1)
    clean.SetTolerance(tol)

# Probably a way to add an observer to the camera position instead
pl.iren.add_observer('MouseWheelForwardEvent', on_zoom)
pl.iren.add_observer('MouseWheelBackwardEvent', on_zoom)
pl.show()
```

I was really lazy figuring out the right formula to set the appropriate tolerance, but it works in principle. Probably worth asking at the [VTK discourse](https://discourse.vtk.org/) if they have better approaches. Maybe there's a way to avoid python callbacks by adding a custom filter after the distanceToCamera filter to set a subset of the 'DistanceToCamera` point scalars to zero so those points get glyphed with a scale factor of zero.
